external/avro/src/test/scala/org/apache/spark/sql/avro/AvroCatalystDataConversionSuite.scala

```diff
@@ -209,4 +209,32 @@ class AvroCatalystDataConversionSuite extends SparkFunSuite
       checkUnsupportedRead(input, avroSchema)
     }
   }
+
+  test("user-specified schema") {
+    val data = Literal("SPADES")
+    val jsonFormatSchema =
+      """
+        |{ "type": "enum",
+        |  "name": "Suit",
+        |  "symbols" : ["SPADES", "HEARTS", "DIAMONDS", "CLUBS"]
+        |}
+      """.stripMargin
+    checkEvaluation(
+      AvroDataToCatalyst(
+        CatalystDataToAvro(
+          data,
+          Some(jsonFormatSchema)),
+        jsonFormatSchema,
+        options = Map.empty),
+      data.eval())
+    val message = intercept[SparkException] {
+      AvroDataToCatalyst(
+        CatalystDataToAvro(
+          data,
+          None),
+        jsonFormatSchema,
+        options = Map.empty).eval()
+    }.getMessage
+    assert(message.contains("Malformed records are detected in record parsing."))
```

Comment:
In this PR, `CatalystDataToAvro` ignores the given scheme in case of `None`, doesn't it? For me, this error seems to come from `AvroDataToCatalyst` instead of `CatalystDataToAvro`.

Reply:
If this error comes from `AvroDataToCatalyst`, this test coverage is misleading. For example, we had better have a test coverage for
- a test whether `CatalystDataToAvro(data, None)` successfully ignores `None` without any exception.
- a test whether `CatalystDataToAvro(data, "")` fails with that error message (?)

How do you think about that, @gengliangwang ?

Reply:
Here `AvroDataToCatalyst` is just to check the Avro schema of `CatalystDataToAvro`.
1. When `jsonFormatSchema` is provided in `CatalystDataToAvro`, the output Avro schema is `enum` type, and we validate it with `AvroDataToCatalyst`. This proves that the provided schema works.
2. When the `jsonFormatSchema` is None, the output Avro schema is `string` type, and it can't be parsed as `enum` type.
I will change the order of the two checks in the case and add a new test case for invalid user-specified schema

Reply:
+1. Thanks, @gengliangwang .